Introduce temporal types to Cypher

thobe opened the pull request:
This introduces temporal `Value` types and the beginning of functions for creating an manipulating them through Cypher.

These functions are defined as built in functions in the database, and not "special cases" in the parser. In order to further support this, the `CallableUserFunction` interface has been updated to work with `Value` objects, moving the conversion to embedded types to the reflective function implementation (that conversion happened on a higher level before, so this means that we can avoid the conversion for non-reflective functions).

thobe committed:
Introduce value types for date and time

thobe committed:
Rename Neo4jValue to DefaultParameterValue

Since that was how it was used.

thobe committed:
CallableUserFunction works with AnyValue.

Introduced a ValueMapper type for this. ValueMapper translates (Any)Value to
other type systems, in this case the java classes of the embedded API.

thobe committed:
Fix rewriter for 'null' default function parameter values

systay commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/TemporalFunctionsAcceptanceTest.scala:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  test("should get current 'now' datetime") {
+    val result = executeWith(supported, "RETURN datetime({when:'now'}) as now")
```
This looks weird to me. I get using maps for mult-argument procedures, to avoid using positional arguments. For single argument procs, I think it's unnecessary to be inconsistent with most other procedures and functions we have and use maps.

thobe replied:
Please see [CIP2015-08-06](https://github.com/thobe/openCypher/blob/date-time/cip/1.accepted/CIP2015-08-06-date-time.adoc), the map is needed here to distinguish from the different overloads of the function. It looks weird because the other tests have not been written yet.

thobe replied:
But your comment got me thinking. So we discussed it and came up with what might be a better way to do this particular thing: https://github.com/opencypher/openCypher/pull/15/commits/5bb14895b95e40a0e3769c9120c9e3b161787227

systay commented:
This makes all kinds of giddy. It's been a long time coming. :)

thobe committed:
Introduce builders for temporal values

thobe committed:
Improve SystemNanoClock

Avoid indirection for millis()

Support withZone(...)

thobe committed:
Introduce ClockContext

thobe committed:
Make it possible to inject a fake clock into the TestDB

thobe committed:
Use ClockContext for injecting clocks into function context

thobe committed:
Introduce temporal functions in Cypher

thobe commented:
I have simplified the content of this branch (in all the commits, to facilitate reviewing), some additional features are moved out into https://github.com/thobe/neo4j/tree/3.4-temporal-experiments

These simplifications are in line with simplifications that have been done to [the specification](https://github.com/thobe/openCypher/blob/date-time/cip/1.accepted/CIP2015-08-06-date-time.adoc). The additional features in that branch are no longer in line with the specification.

thobe committed:
CastSupport for temporal values

thobe committed:
Support ValueWriter from temporal values

thobe committed:
Don't make equals final

Mockito barfs on this.

Really we should not be mocking values, but that's a change for another time.

thobe committed:
Do not rely on order of keys in maps

thobe commented:
I'm not sure the commit entitled `CastSupport for temporal values` is actually needed, it just follows a pattern that existed in that file.

thobe committed:
Introduce all remaining temporal functions

Also a slight simpligying refactoring to StructureBuilder.

thobe commented:
Remaining work:

- [ ] Lots of testing needs to be added.
- [ ] There are quite a few blanks left to fill in for the implementation of the temporal functions.
- [ ] Storage of temporal values. _After which we can:_
- [ ] Add tests to the Core API for storing and retrieving `java.time.{ZonedDateTime,LocalDateTime,LocalDate,OffsetTime,LocalTime}`
- [ ] BOLT serialisation of temporal values.
- [ ] Support for reading temporal values from CSV files in the importer (by using the static `parse(...)` methods of the respective types)

sherfert commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/PathProxy.java:
```diff
@@ -0,0 +1,284 @@
+                }
+                catch ( Exception e )
+                {
+                    inTx = false;
```
Is there another way to figure out, if we are in a trnasaction currently?

thobe replied:
Through quite a bit of refactoring.

thobe replied:
There is a bit of redemption though, which makes this solution not-so-bad, and makes that refactoring not-really-worth-it, and that is that typically the `getRelationshipTypeById(...)`-method works even outside of a transaction - although it is not guaranteed to... (and would definitely fail if the database has been shut down)

sherfert commented on community/values/src/main/java/org/neo4j/values/storable/DurationValue.java:
```diff
@@ -0,0 +1,805 @@
+            {
+                throw new IllegalArgumentException( "hours out of range: " + h );
+            }
+            if ( m > 60 )
```
I could only find the  valid ranges for minute, second, etc. for the `Time` types. If the caps are at 59 for minutes and 60 for seconds _there_, why are they both at 60 for durations?

thobe replied:
good point, should have been `>=`

thobe replied:
Classic kind of thing that should be covered by a test case :)

thobe commented:
@sherfert I actually think that writing tests for some code is a great way of reviewing it, so in a way the low test coverage is a feature of the pull request...

sherfert commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -168,9 +169,9 @@ else if ( this.coordinate.length < other.coordinate.length )
+    public Point asObjectCopy()
     {
-        return new PointValue( this.crs, this.coordinate );
+        return this;
```
I am unsure if it is OK not to create a new object (as done with `clone` in other implementations of `asObjectCopy`).

thobe replied:
If you read the rest of the code, you will find that `PointValue` is never mutated, and that the `coordinate` field is not made public to anything that can mutate it.

Would applying this diff make you more comfortable?

```
--- a/community/values/src/main/java/org/neo4j/values/storable/PointValue.java
+++ b/community/values/src/main/java/org/neo4j/values/storable/PointValue.java
@@ -32,8 +32,8 @@ import static java.util.Collections.singletonList;
 
 public class PointValue extends ScalarValue implements Comparable<PointValue>, Point
 {
-    private CoordinateReferenceSystem crs;
-    private double[] coordinate;
+    private final CoordinateReferenceSystem crs;
+    private final double[] coordinate;
 
     PointValue( CoordinateReferenceSystem crs, double... coordinate )
     {
```

thobe replied:
Looking further at the usages of the `coordinate()` method through which the `double[] coordinate` field might leak, I think it would make a lot of sense to instead expose these two methods:

```
public int numberOfCoordinates()
{
    return coordinate.length;
}
public double coordinate( int i )
{
    return coordinate[i];
}
```

sherfert commented on community/cypher/runtime-util/src/test/scala/org/neo4j/cypher/FakeClock.scala:
```diff
@@ -0,0 +1,28 @@
+
+import org.neo4j.time.Clocks
+
+trait FakeClock {
```
Would it make sense to add this here?
```
self: GraphDatabaseTestSupport  =>
```

thobe replied:
Maybe. It wasn't needed, so I didn't bother with the extra typing.

sherfert committed:
Extending DateTimeValueTest

sherfert committed:
More tests, including for add and sub.

Adding toString to DurationValue.

sherfert committed:
Fix checkstyle errors.

thobe committed:
Disallow invalid combinations when constructing time

If combining 'millisecond', 'microsecond', and 'nanosecond' with one
another require that each is in the range of [0,1000[, if only one
of them is passed allone, it can have its full range.

henriknyman reviewed (approved):
Looks good so far. 
I guess we want to also consider changing procedure call in a similar way as function call?
Have you benchmarked any performance impact?

thobe commented:
@henriknyman yes, I agree that we would want to change procedures similarly to how functions are changed here, but I wanted to limit the scope of this particular changeset.

I have not yet measured the performance impact, but I have added it to the benchmarking queue, so it should be done soon.

sherfert committed:
Add more test coverage for DateTimeValueBuilder
